contracts/staking/StakingBase.sol

```diff
@@ -661,17 +774,52 @@ abstract contract StakingBase is ERC721TokenReceiver {
                 lastAvailableRewards += reward;
                 availableRewards = lastAvailableRewards;
             } else {
-                _withdraw(multisig, reward);
+                // Get reward receivers and amounts based on the distribution type
+                (address[] memory receivers, uint256[] memory amounts) =
+                    _getRewardReceiversAndAmounts(serviceId, sInfo.owner, sInfo.multisig, sInfo.reward, distributionType);
+
+                // Transfer reward amounts to specified receivers
+                _withdraw(receivers, amounts);
             }
         }
 
+        // Transfer the service back to the owner
+        // Note that the reentrancy is not possible due to the ServiceInfo struct being deleted
+        IService(serviceRegistry).transferFrom(address(this), msg.sender, serviceId);
+
         if (enforced) {
-            emit ServiceForceUnstaked(epochCounter, serviceId, msg.sender, multisig, nonces, reward, lastAvailableRewards);
+            emit ServiceForceUnstaked(epochCounter, serviceId, msg.sender, sInfo.multisig, sInfo.nonces, sInfo.reward,
+                lastAvailableRewards);
         } else {
-            emit ServiceUnstaked(epochCounter, serviceId, msg.sender, multisig, nonces, reward, lastAvailableRewards);
+            emit ServiceUnstaked(epochCounter, serviceId, msg.sender, sInfo.multisig, sInfo.nonces, sInfo.reward,
+                lastAvailableRewards);
         }
     }
 
+    /// @dev Withdraws reward amounts.
+    /// @notice The balance is always greater or equal the amount, as follows from the Base contract logic.
+    /// @param receivers Set of receiver addresses.
+    /// @param amounts Corresponding amounts to withdraw.
+    function _withdraw(address[] memory receivers, uint256[] memory amounts) internal virtual {
```

Comment:
Withdraw was a copy-paste in each `StakingNativeToken` and `StakingToken` contracts, whereas follows the same logic. The asset transfer is the only difference => thus the `_transfer()` fully virtual function overridden in each contract.